lib/core/src/text/content.rs

```diff
@@ -0,0 +1,370 @@
+
+/// Set of dirty lines' indices
+#[derive(Debug)]
+pub struct DirtyLines {
+    pub single_lines : HashSet<usize>,
+    pub range        : Option<RangeFrom<usize>>
+}
+
+impl Default for DirtyLines {
+    /// Default `DirtyLines` where no line is dirty.
+    fn default() -> Self {
+        Self {
+            single_lines : HashSet::new(),
+            range        : None,
+        }
+    }
+}
+
+impl DirtyLines {
+    /// Mark single line as dirty.
+    pub fn add_single_line(&mut self, index:usize) {
+        self.single_lines.insert(index);
+    }
+
+    /// Mark an open range of lines as dirty.
+    pub fn add_lines_range_from(&mut self, range:RangeFrom<usize>) {
+        let current_is_wider = self.range.as_ref().map_or(false, |cr| cr.start <= range.start);
+        if !current_is_wider {
+            self.range = Some(range);
+        }
+    }
+
+    /// Mark an open range of lines as dirty.
+    pub fn add_lines_range(&mut self, range:RangeInclusive<usize>) {
+        for i in range {
+            self.add_single_line(i);
+        }
+    }
+
+    /// Check if line is marked as dirty.
+    pub fn is_dirty(&self, index:usize) -> bool {
+        let range_contains = self.range.as_ref().map_or(false, |r| r.contains(&index));
+        range_contains || self.single_lines.contains(&index)
+    }
+
+    /// Check if there is any dirty line
+    pub fn any_dirty(&self) -> bool {
+        self.range.is_some() || !self.single_lines.is_empty()
+    }
+}
+
+// ==============
+// === Change ===
+// ==============
+
+/// A change type
+///
+/// A change is simple if it's replace a fragment of one line with text without newlines. Otherwise
+/// its a multiline change.
+pub enum ChangeType {
+    Simple, Multiline
+}
+
+/// A structure describing a text operation in one place.
+pub struct TextChange {
+    replaced : Range<CharPosition>,
+    lines    : Vec<String>,
+}
+
+impl TextChange {
+    /// Creates operation which inserts text at given position.
+    pub fn insert(position:CharPosition, text:&str) -> Self {
+        TextChange {
+            replaced : position.clone()..position,
+            lines    : TextComponentContent::split_to_lines(text)
+        }
+    }
+
+    /// Creates operation which deletes text at given range.
+    pub fn delete(range:Range<CharPosition>) -> Self {
+        TextChange {
+            replaced : range,
+            lines    : vec!["".to_string()],
+        }
+    }
+
+    /// Creates operation which replaces text at given range with given string.
+    pub fn replace(replaced:Range<CharPosition>, text:&str) -> Self {
+        TextChange {replaced,
+            lines : TextComponentContent::split_to_lines(text)
+        }
+    }
+
+    /// A type of this change. See `ChangeType` doc for details.
+    pub fn change_type(&self) -> ChangeType {
+        if self.lines.is_empty() {
+            panic!("Invalid change");
+        } else if self.replaced.start.line == self.replaced.end.line && self.lines.len() == 1 {
```

Comment:
Its hard to undersstand if the condition is correct and why its working the way its working. I would consider refactoring it somehow.